circuit/types/field/src/div_flagged.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+impl<E: Environment> DivFlagged<Field<E>> for Field<E> {
+    type Output = (Field<E>, Boolean<E>);
+
+    /// Performs flagged division of two field elements.
+    #[doc(hidden)]
+    fn div_flagged(&self, other: &Field<E>) -> Self::Output {
+        match (self.is_constant(), other.is_constant()) {
+            // If `other` is a constant and zero, return the zero element and set the flag.
+            (_, true) if other.eject_value().is_zero() => (Field::zero(), Boolean::constant(true)),
+            // If `other` is a constant and non-zero, we can perform multiplication and inversion for 0 constraints.
+            // If `self` is a constant, we can perform multiplication and inversion for 1 constraint.
+            // Otherwise, we can perform multiplication and inversion for 2 constraints.
+            _ => {
+                let (other_inverse, flag) = other.inverse_flagged();
+                (self * other_inverse, flag)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use snarkvm_circuit_environment::Circuit;
+
+    const ITERATIONS: u64 = 1000;
+
+    fn check_div_flagged(
+        name: &str,
+        first: &console::Field<<Circuit as Environment>::Network>,
+        second: &console::Field<<Circuit as Environment>::Network>,
+        mode_a: Mode,
+        mode_b: Mode,
+    ) {
+        let a = &Field::<Circuit>::new(mode_a, *first);
+        let b = &Field::<Circuit>::new(mode_b, *second);
+
+        match second.is_zero() {
+            true => {
+                // Check that `div_flagged` produces the correct result and that the flag is set.
+                Circuit::scope(name, || {
+                    let (result, flag) = a.div_flagged(b);
+                    assert_eq!(result.eject_value(), console::Field::zero());
+                    assert!(flag.eject_value());
+                });
+                Circuit::reset();
+
+                // Check that `div` produces the correct result.
+                match mode_b.is_constant() {
+                    true => {
+                        // Check that `div` halts.
+                        let result = std::panic::catch_unwind(|| a.div(b));
```

Comment:
Why does that initially panic instead of !bailing into an Error monad? I know panicking is often the right decision in many contexts (like within SGX, etc.) - just wondering why specifically here.

Reply:
I'm guessing it's due to halting; to change that one to not contain a panic would be a huge undertaking